src/java.base/share/classes/sun/security/util/PBKDF2Parameters.java

```diff
@@ -0,0 +1,211 @@
+     * parameter block.
+     *
+     * @param pBKDF2_params the DER encoding of the parameter block
+     *
+     * @throws IOException for parsing errors in the input stream
+     */
+    public PBKDF2Parameters(DerValue pBKDF2_params) throws IOException {
+
+        if (pBKDF2_params.tag != DerValue.tag_Sequence) {
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue specified = pBKDF2_params.data.getDerValue();
+        // the 'specified' ASN.1 CHOICE for 'salt' is supported
+        if (specified.tag == DerValue.tag_OctetString) {
+            salt = specified.getOctetString();
+        } else {
+            // the 'otherSource' ASN.1 CHOICE for 'salt' is not supported
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "not an ASN.1 OCTET STRING tag");
+        }
+        iterationCount = pBKDF2_params.data.getInteger();
+
+        // keyLength INTEGER (1..MAX) OPTIONAL,
+        var ksDer = pBKDF2_params.data.getOptional(DerValue.tag_Integer);
+        if (ksDer.isPresent()) {
+            keyLength = ksDer.get().getInteger() * 8; // keyLength (in bits)
+        } else {
+            keyLength = -1;
+        }
+
+        // prf AlgorithmIdentifier {{PBKDF2-PRFs}} DEFAULT algid-hmacWithSHA1
+        var prfDer = pBKDF2_params.data.getOptional(DerValue.tag_Sequence);
+        if (prfDer.isPresent()) {
+            DerValue prf = prfDer.get();
+            // the pseudorandom function (default is HmacSHA1)
+            ObjectIdentifier kdfAlgo_OID = prf.data.getOID();
+            KnownOIDs o = KnownOIDs.findMatch(kdfAlgo_OID.toString());
+            if (o == null || (!o.stdName().equals("HmacSHA1") &&
+                    !o.stdName().equals("HmacSHA224") &&
+                    !o.stdName().equals("HmacSHA256") &&
+                    !o.stdName().equals("HmacSHA384") &&
+                    !o.stdName().equals("HmacSHA512") &&
+                    !o.stdName().equals("HmacSHA512/224") &&
+                    !o.stdName().equals("HmacSHA512/256"))) {
+                throw new IOException("PBKDF2 parameter parsing error: "
+                        + "expecting the object identifier for a HmacSHA "
+                        + "pseudorandom function");
+            }
+            prfAlgo = o.stdName();
+            prf.data.getOptional(DerValue.tag_Null);
+            prf.data.atEnd();
+        } else {
+            prfAlgo = "HmacSHA1";
+        }
+    }
+
+    public static byte[] encode(byte[] salt, int iterationCount,
+            int keyLength, String kdfHmac) {
+        ObjectIdentifier prf =
+               ObjectIdentifier.of(KnownOIDs.findMatch(kdfHmac));
+        return PBKDF2Parameters.encode(salt, iterationCount, keyLength, prf);
+    }
+
+    /*
+     * Encode PBKDF2 parameters from components.
+     */
+    public static byte[] encode(byte[] salt, int iterationCount,
```

Comment:
This method actually encodes more than the parameters. It also encodes the outer algorithm id. I guess that's ok, but I suggest adding a comment that says that.

Reply:
added